### PR TITLE
Improve metrics hooks setup (fixes #12672)

### LINKED ISSUE #12672: Improve metrics hook setup
### Describe the feature

currently we do:

https://github.com/paradigmxyz/reth/blob/3408059393bcf03f6727f790ec52f28114e25d02/crates/node/metrics/src/hooks.rs#L24-L26

which makes the metrics crate depend on 

https://github.com/paradigmxyz/reth/blob/3408059393bcf03f6727f790ec52f28114e25d02/crates/node/metrics/src/hooks.rs#L4-L4

we can improve this api with a builder API

```rust
HooksBuilder {
  Vec<Box<dyn Hook<Output = ()>>>
}
```

and replace the new function 

https://github.com/paradigmxyz/reth/blob/3408059393bcf03f6727f790ec52f28114e25d02/crates/node/metrics/src/hooks.rs#L30-L30

with something like

```
fn with(mut self, hook: impl Hook)
```

then we need to change how this type is instantiated and effectively push these:

https://github.com/paradigmxyz/reth/blob/3408059393bcf03f6727f790ec52f28114e25d02/crates/node/metrics/src/hooks.rs#L36-L41

to the callsite

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nils-mathieu**
This pull request fixes #12672.

It adds the type `HooksBuilder`, responsible for creating new custom instances of the `Hooks` collection.

# Changes

- Add the `HooksBuilder` type.
- Add the `Hooks::builder` method to create a new `HooksBuilder`.
- Change the calls to `Hooks::new` by `Hooks::builder`.
- Remove `Hooks::new`.

# Considerations

- I added some convenience methods to `HooksBuilder` to ease its usage a bit.
  - `HooksBuilder::with_database_metrics`
  - `HooksBuilder::with_static_file_provider`
  - `HooksBuilder::with_default_collector`

I made that choice because moving the cloning logic to the caller of `Hooks::new` requires us to write a bit of boilerplate due to having to move resources into the static hook closure:

```rust
    .with_hook({
        let db = factory.db_ref().clone();
        move || db.report_metrics()
    })
```

Replaced by:

```rust
    .with_database_metrics(factory.db_ref().clone())
```

I'm not sure whether this was a good idea. It is a bit faster to write, but maybe it makes the logic a bit harder to reason about. If someone reads that, they might assume that `with_database_metrics` is setting a specific "database_metrics" field in the builder, when it is not.

I can remove those convenience methods if needed.

# Other

- I had to fix a compilation error here (there was a comparaison between `i32` and `u32`:
https://github.com/paradigmxyz/reth/blob/7c7baca9807e9a9556cb217d3df8cd03c21ba4ed/crates/storage/libmdbx-rs/src/environment.rs#L431-L444

- I changed a `format!` to a `format_args!` in the `Debug` implementation of `Hooks` to save on an allocation.

### CODE REVIEW

**Comment by nils-mathieu** [comment]
The `ffi::MDBX_RDONLY` type and friends seem to have different types on my machine (Windows 11) than they are on the testing environment.

Should I duplicate the function? Ideally the contant themselves should change not to change type depending on the compilation target, I guess.

**Review by mattsse — CHANGES_REQUESTED** [review]
great start

**Comment by mattsse** [comment]
regarding the different mdbx type, could you ptal at

https://github.com/paradigmxyz/reth/issues/12259#issuecomment-2452962847

File: crates/node/metrics/src/hooks.rs [hunk 1]
```diff
@@ -2,19 +2,78 @@ use metrics_process::Collector;
...
+        self,
+        metrics: impl 'static + Send + Sync + DatabaseMetrics,
+    ) -> Self {
+        self.with_hook(move || metrics.report_metrics())
+    }
+
+    /// Registers a hook that uses a [`NodePrimitives`] instance to invoke.
+    pub fn with_static_file_provider<N: NodePrimitives>(
+        self,
+        provider: StaticFileProvider<N>,
```
[the remark this answers is not in the document]
Comment (nils-mathieu) [reply crates/node/metrics/src/hooks.rs:41]: understood, will do

File: crates/node/metrics/Cargo.toml [hunk 2]
```diff
@@ -8,9 +8,6 @@ homepage.workspace = true
 repository.workspace = true
 
 [dependencies]
-reth-db-api.workspace = true
-reth-primitives-traits.workspace = true
-reth-provider.workspace = true
```
Comment (mattsse) [inline crates/node/metrics/Cargo.toml:13]: great

**Review by mattsse — COMMENTED** [review]
lgtm, 

could you please apply the changes here

https://github.com/paradigmxyz/reth/pull/12698

can't push to your main branch

**Comment by nils-mathieu** [comment]
@mattsse done!

**Review by mattsse — APPROVED** [review]

**Pull request merged**